### PR TITLE
[Merged by Bors] - feat: port Archive.Wiedijk100Theorems.AscendingDescendingSequences

### PR DESCRIPTION
**Pull request by Komyyy**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean [hunk 1]
```diff
@@ -138,15 +138,15 @@ theorem erdos_szekeres {r s n : ℕ} {f : Fin n → α} (hn : r * s < n) (hf : I
   -- Now that we have uniqueness of each label, it remains to do some counting to finish off.
   -- Suppose all the labels are small.
   by_contra q
-  push_neg at q 
+  simp (config := { zeta := false, proj := false }) only [not_exists, not_or, Nat.not_lt] at q
```
Comment (semorrison) [inline Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean]: This needs a porting note quoting the original tactic, and noting a regression in `push_neg`.

File: Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean [hunk 2]
```diff
@@ -138,15 +138,15 @@ theorem erdos_szekeres {r s n : ℕ} {f : Fin n → α} (hn : r * s < n) (hf : I
...
   -- Then the labels `(a_i, b_i)` all fit in the following set: `{ (x,y) | 1 ≤ x ≤ r, 1 ≤ y ≤ s }`
   let ran : Finset (ℕ × ℕ) := (range r).image Nat.succ ×ˢ (range s).image Nat.succ
   -- which we prove here.
   have : image ab univ ⊆ ran := by
     -- First some logical shuffling
     rintro ⟨x₁, x₂⟩
-    simp only [mem_image, exists_prop, mem_range, mem_univ, mem_product, true_and_iff,
-      Prod.mk.inj_iff]
+    simp (config := { zeta := false, proj := false }) only [mem_image, exists_prop, mem_range,
+      mem_univ, mem_product, true_and_iff, Prod.ext_iff]
```
Comment (semorrison) [inline Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean:149]: Porting note? Why the custom config?

File: Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean [hunk 3]
```diff
@@ -158,17 +158,17 @@ theorem erdos_szekeres {r s n : ℕ} {f : Fin n → α} (hn : r * s < n) (hf : I
           refine' ⟨{i}, by solve_by_elim, card_singleton i⟩
     refine' ⟨_, _⟩
     -- Need to get `a_i ≤ r`, here phrased as: there is some `a < r` with `a+1 = a_i`.
-    · refine' ⟨(ab i).1 - 1, _, Nat.succ_pred_eq_of_pos z.1⟩
-      rw [tsub_lt_iff_right z.1]
+    · refine' ⟨Nat.pred (ab i).1, _, Nat.succ_pred_eq_of_pos z.1⟩
+      rw [Nat.pred_eq_sub_one, tsub_lt_iff_right z.1]
```
Comment (semorrison) [inline Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean:162]: What's going on here?

File: Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean [hunk 4]
```diff
@@ -158,17 +158,17 @@ theorem erdos_szekeres {r s n : ℕ} {f : Fin n → α} (hn : r * s < n) (hf : I
...
-    · refine' ⟨(ab i).2 - 1, _, Nat.succ_pred_eq_of_pos z.2⟩
-      rw [tsub_lt_iff_right z.2]
+    · refine' ⟨Nat.pred (ab i).2, _, Nat.succ_pred_eq_of_pos z.2⟩
+      rw [Nat.pred_eq_sub_one, tsub_lt_iff_right z.2]
       apply Nat.lt_succ_of_le q.2
   -- To get our contradiction, it suffices to prove `n ≤ r * s`
   apply not_le_of_lt hn
   -- Which follows from considering the cardinalities of the subset above, since `ab` is injective.
-  simpa [Nat.succ_injective, card_image_of_injective, ‹injective ab›] using card_le_of_subset this
+  simpa (config := { zeta := false }) [Nat.succ_injective, card_image_of_injective, ‹Injective ab›]
```
Comment (semorrison) [inline Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean:170]: Why the custom config?

**Comment by Komyyy** [comment]
I've obscured the definition of `ab` so these workarounds and additional `maxHeartbeats` are unnecesarry anymore.

File: Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean [hunk 5]
```diff
@@ -64,35 +63,38 @@ theorem erdos_szekeres {r s n : ℕ} {f : Fin n → α} (hn : r * s < n) (hf : I
...
   -- We call these labels `(a_i, b_i)`.
-  let ab : Fin n → ℕ × ℕ := by
+  let ab' : Fin n → ℕ × ℕ := by
     intro i
     apply
-      (max' ((inc_sequences_ending_in i).image card) (nonempty.image ⟨{i}, inc_i i⟩ _),
-        max' ((dec_sequences_ending_in i).image card) (nonempty.image ⟨{i}, dec_i i⟩ _))
+      (max' ((inc_sequences_ending_in i).image card) (Nonempty.image ⟨{i}, inc_i i⟩ _),
+        max' ((dec_sequences_ending_in i).image card) (Nonempty.image ⟨{i}, dec_i i⟩ _))
+  -- Porting note: it costs many resources to unfold this function so we obscure the definition.
```
Comment (jcommelin) [inline Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean:71]:
```suggestion
  -- Porting note: it costs many resources to unfold `ab'` so we obscure the definition:
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**